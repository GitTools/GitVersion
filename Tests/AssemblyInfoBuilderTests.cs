using System;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using ApprovalTests;
using GitVersion;
using GitVersionTask;
using NUnit.Framework;
using Roslyn.Compilers;
using Roslyn.Compilers.CSharp;

[TestFixture]
public class AssemblyInfoBuilderTests
{

    [Test]
    public void VerifyCreatedCode()
    {
        var semanticVersion = new SemanticVersion
        {
            Major = 1,
            Minor = 2,
            Patch = 3,
            PreReleaseTag = "unstable4",
            BuildMetaData = new SemanticVersionBuildMetaData(5,
                "feature-Add_xyz",
                new ReleaseDate
                {
                    OriginalCommitSha = "originalCommitSha",
                    OriginalDate = DateTimeOffset.Parse("2014-03-01 00:00:01Z"),
                    CommitSha = "commitSha",
                    Date = DateTimeOffset.Parse("2014-03-06 23:59:59Z")
                })
        };
        var assemblyInfoBuilder = new AssemblyInfoBuilder
            {
                SemanticVersion = semanticVersion
            };
        var assemblyInfoText = assemblyInfoBuilder.GetAssemblyInfoText();
        Approvals.Verify(assemblyInfoText);
        var syntaxTree = SyntaxTree.ParseText(assemblyInfoText);
        var references = new[] {new MetadataFileReference(typeof(object).Assembly.Location), };
        var compilation = Compilation.Create("Greeter.dll", new CompilationOptions(OutputKind.NetModule), new[] { syntaxTree }, references);
        var emitResult = compilation.Emit(new MemoryStream());
        Assert.IsTrue(emitResult.Success, string.Join(Environment.NewLine, emitResult.Diagnostics.Select(x => x.Info)));
    }

    [Test]
    [MethodImpl(MethodImplOptions.NoInlining)]
    public void VerifyAssemblyVersion_Major()
    {
        VerifyAssemblyVersion(AssemblyVersioningScheme.Major);
    }

    [Test]
    [MethodImpl(MethodImplOptions.NoInlining)]
    public void VerifyAssemblyVersion_MajorMinor()
    {
        VerifyAssemblyVersion(AssemblyVersioningScheme.MajorMinor);
    }

    [Test]
    [MethodImpl(MethodImplOptions.NoInlining)]
    public void VerifyAssemblyVersion_MajorMinorPatch()
    {
        VerifyAssemblyVersion(AssemblyVersioningScheme.MajorMinorPatch);
    }

    [Test]
    [MethodImpl(MethodImplOptions.NoInlining)]
    public void VerifyAssemblyVersion_None()
    {
        VerifyAssemblyVersion(AssemblyVersioningScheme.None);
    }

    static void VerifyAssemblyVersion(AssemblyVersioningScheme avs)
    {
        var semanticVersion = new SemanticVersion
        {
            Major = 2,
            Minor = 3,
            Patch = 4,
            BuildMetaData = new SemanticVersionBuildMetaData(5,
                "master",
                new ReleaseDate
                {
                    OriginalCommitSha = "originalCommitSha",
                    OriginalDate = DateTimeOffset.Parse("2014-03-01 00:00:01Z"),
                    CommitSha = "commitSha",
                    Date = DateTimeOffset.Parse("2014-03-06 23:59:59Z")
                }),
        };
        var assemblyInfoBuilder = new AssemblyInfoBuilder
        {
            SemanticVersion = semanticVersion,
            AssemblyVersioningScheme = avs,
        };

        var assemblyInfoText = assemblyInfoBuilder.GetAssemblyInfoText();
        Approvals.Verify(assemblyInfoText);
        var syntaxTree = SyntaxTree.ParseText(assemblyInfoText);
        var references = new[] { new MetadataFileReference(typeof(object).Assembly.Location), };
        var compilation = Compilation.Create("Greeter.dll", new CompilationOptions(OutputKind.NetModule), new[] { syntaxTree }, references);
        var emitResult = compilation.Emit(new MemoryStream());
        Assert.IsTrue(emitResult.Success, string.Join(Environment.NewLine, emitResult.Diagnostics.Select(x => x.Info)));
    }
}
