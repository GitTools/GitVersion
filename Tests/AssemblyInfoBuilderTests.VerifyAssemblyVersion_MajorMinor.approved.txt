
using System;
using System.Reflection;

[assembly: AssemblyVersion("2.3.0.0")]
[assembly: AssemblyFileVersion("2.3.4.0")]
[assembly: AssemblyInformationalVersion("2.3.4+5.Branch.master.Sha.commitSha")]
[assembly: ReleaseDate("2014-03-01", "2014-03-06")]

[System.Runtime.CompilerServices.CompilerGenerated]
sealed class ReleaseDateAttribute : System.Attribute
{
    public string OriginalDate { get; private set; }
    public string Date { get; private set; }

    public ReleaseDateAttribute(string originalDate, string date)
    {
        OriginalDate = date;
        Date = date;
    }
}

[System.Runtime.CompilerServices.CompilerGenerated]
static class GitVersionInformation
{
    public static string Major = "2";
    public static string Minor = "3";
    public static string Patch = "4";
    public static string PreReleaseTag = "";
    public static string PreReleaseTagWithDash = "";
    public static string BuildMetaData = "5";
    public static string FullBuildMetaData = "5.Branch.master.Sha.commitSha";
    public static string MajorMinorPatch = "2.3.4";
    public static string SemVer = "2.3.4";
    public static string LegacySemVer = "2.3.4";
    public static string LegacySemVerPadded = "2.3.4";
    public static string AssemblySemVer = "2.3.4.0";
    public static string FullSemVer = "2.3.4+5";
    public static string InformationalVersion = "2.3.4+5.Branch.master.Sha.commitSha";
    public static string ClassicVersion = "2.3.4.5";
    public static string ClassicVersionWithTag = "2.3.4.5";
    public static string BranchName = "master";
    public static string Sha = "commitSha";
    public static string AssemblyVersion = "2.3.0.0";
    public static string AssemblyFileVersion = "2.3.4.0";
    public static string OriginalRelease = "originalCommitSha.2014-03-01 00:00:01Z";
    public static string NugetPackageVersion = "2.3.4";

}


